Add Pony for Python Programmers write up

KushalP opened the pull request:
Going to submit a PR for this early to get some feedback. I'm not happy with the structure and I really want to get some concurrency bits in at some point.

It's very similar to ponylang/pony-for-x#4

KushalP committed:
Add Pony for Python Programmers write up

SeanTAllen commented:
@aturley can you take a look at this? i'm not really qualified Python wise (as my cursing when writing "modern Python" attests to)

SeanTAllen commented on python.asciidoc:
```diff
@@ -0,0 +1,564 @@
+consistently used. Pony uses distinct keywords, with `var` to introduce
+variables, and `let` to introduce values
+
+[[installing-pony]]
```
same installing comment I made in the ruby version:

we've already had issue maintaining directions etc in different places.

this should be replaced with a link to the install instructions in the main repo README.

KushalP replied:
Updated in 94322a5.

KushalP committed:
Add concurrenct hello world for Python

aturley commented on python.asciidoc:
```diff
@@ -0,0 +1,562 @@
+$ ./testing
+1 test started, 0 complete: addition started
+1 test started, 1 complete: addition complete
+---- Passed: addition
```
The dashed lines from the output are screwing up the formatting because the match the dashed lines that delimit the source block. This also happens in a few places below.

aturley commented on python.asciidoc:
```diff
@@ -0,0 +1,562 @@
+    | if a > b => "a is bigger than b"
+    | if a < b => "b is bigger than a"
+    else "a and b are the same"
+  end
```
The `end` here should line up with the `match` since it is ending the `match` block.

aturley commented on python.asciidoc:
```diff
@@ -0,0 +1,562 @@
+    def dimensions(self):
+        return "{w} {h}".format(w=self._width, h=self._height)
+
+    def circumference(self):
```
Not really a Python/Pony issue, but I'm not sure `circumference` is the right function name here. "Circumference" is most commonly used to refer to perimeter of ellipses (and, by extension, circles), not rectangles. I see that the Java chapter uses the same name, and I'd rather fix it in both places than propagate the error.

aturley commented on python.asciidoc:
```diff
@@ -0,0 +1,562 @@
+[[getting-functional]]
+Getting functional
+~~~~~~~~~~~~~~~~~~
+
```
I have a quibble with this example. Pony has an `elseif` keyword, so the Pony equivalent of the Python code would probably use that.

```
    if a > b then
      "a is bigger than b"
    elseif a < b => then
      "b is bigger than a"
    else 
      "a and b are the same"
    end
```

I'd argue that the `match` statement shown in this example isn't really representative of the way that `match` is used in most Pony code. At the very least, I think the Pony `match` statement more closely corresponds to the pattern of [dictionary mapping](http://www.pydanny.com/why-doesnt-python-have-switch-case.html) that you see in Python programs.

I see that this is similar to the example from the Java chapter, and I have the same quibble there.

KushalP replied:
Amended in e117dc3.

aturley commented:
I think this is on the right general path.

This seems to borrow it's structure from the Java chapter. I actually think that Pony and Python have a number of things in common that Java lacks, and I think it would be good to focus on those things. For example:
- Both languages have anonymous functions (lambdas). How are these similar? How are they different?
- [Structural subtyping](https://tutorial.ponylang.org/types/traits-and-interfaces.html) (Pony) can be thought of as a type-safe version of duck-typing (Python). Duck typing says, "if it walks like a duck and it quacks like a duck then it must be a duck", whereas structural subtyping says, "a duck walks like this and quacks like this, so if something does both of those things then it must be a duck."
- Both languages have tuples. How are they similar, and how are they different? If nothing else, both offer tuple destructuring, which could be a good topic of discussion.
- In Python you can pass functions around by name, which is very useful. Pony doesn't let you do that but you can achieve something similar by creating a partial function application and passing that around.
- Python has standalone functions, whereas all Pony functions are part of classes or primitives. Are there any differences in how you might approach solving problems based on this?
- Python programmers are used to having a number of very useful data structures available to them, such as dictionaries, lists, and sets; in fact I've worked with Python developers who use the data structures rather than creating new classes unless they absolutely need a class. Pony's standard library provides these data structures. How might the use of these data structures differ between the languages?
- Python doesn't have a `switch` statement like other languages have, and Python developers have come up with a number of ways to work around this (dictionary mappings, lookup lists, etc). Pony provides a `match` statement which is similar to `switch` statements in other languages. How might the `match` statement provide solutions to problems that might have been solved in Python with the aforementioned techniques?
- Python and Pony both have classes and inheritance. How are they similar? How are they different? How do traits and interfaces fit into that picture?

I'm not saying you need to tackle all of these issues, but I think they would provide a good bridge for a Python programmer to begin to understand what Pony has to offer.

KushalP committed:
Use the explicit `if` form over `match`

This is similar to the Python equivalent and this isn't the best
use-case for `match`.

KushalP commented:
@aturley This is brilliant feedback! I think I might convert this list into a bunch of TODO items.

aturley commented:
@KushalP yeah, @SeanTAllen and I realized there was a need for it. I added a section to the CONTRIBUTING.md file that I'll PR in a bit: [aturley's CONTRIBUTING.md](https://github.com/aturley/pony-for-x/blob/how-to-compare-pony-to-language-x/CONTRIBUTING.md#how-to-compare-pony-to-language-x)
